primitives/src/block.rs

```diff
@@ -201,6 +201,74 @@ impl Encodable for Block {
     }
 }
 
+#[cfg(feature = "alloc")]
+type BlockInnerDecoder = Decoder2<HeaderDecoder, VecDecoder<Transaction>>;
+
+/// The decoder for the [`Block`] type.
+///
+/// This decoder can only produce a `Block<Unchecked>`.
+#[cfg(feature = "alloc")]
+pub struct BlockDecoder(BlockInnerDecoder);
+
+#[cfg(feature = "alloc")]
+impl Decoder for BlockDecoder {
+    type Output = Block;
+    type Error = BlockDecoderError;
+
+    fn push_bytes(&mut self, bytes: &mut &[u8]) -> Result<bool, Self::Error> {
+        self.0.push_bytes(bytes).map_err(BlockDecoderError)
+    }
+
+    fn read_limit(&self) -> usize { self.0.read_limit() }
+
+    fn end(self) -> Result<Self::Output, Self::Error> {
+        let (header, transactions) = self.0.end().map_err(BlockDecoderError)?;
+        Ok(Self::Output::new_unchecked(header, transactions))
+    }
+}
```

Comment:
nit: Can you put the functions in the same order as in the trait please (move `end` above `read_limit`). Also we can inline all three.